Puzzles V3

Endaris opened the pull request:
Summary

  This PR represents a major overhaul and enhancement of the puzzle system in Panel Attack,
  introducing comprehensive improvements to puzzle management, UI/UX, and gameplay mechanics.
   The changes include a complete puzzle editor, enhanced menu system, better user feedback,
  and significant code quality improvements.

  Key Features Added

  • Complete Puzzle Editor - New visual puzzle editor with cursor-based editing, garbage
  drawing, and real-time validation, you can not currently edit the puzzle type or add new puzzles
  • Enhanced Puzzle Menu System - Hierarchical puzzle navigation with completion tracking,
  hints display, and improved visual feedback
  • Puzzle Progress Tracking - Visual indicators for completion status (complete, partial,
  hint availability)
  • Improved Localization - Added translation for puzzle-related UI elements
  • Better User Experience - Consistent menu widths, spacing improvements, and informative
  descriptions
  • Performance Optimizations - Enhanced test launcher with filtering

  Technical Improvements

  • New UI Components - PuzzleGoalDisplay, PuzzleHierarchyDisplay, MultibarElement,
  ImageButton, and more
  • Enhanced Testing Infrastructure - filter tests, improved test launcher with
   exit codes
  • Puzzle Format Evolution - Support for puzzle format V3 with new timing behaviors and
  validation

  Major New Components:
  - client/src/scenes/PuzzleEditorScene.lua  - Complete puzzle editor
  implementation
  - client/src/PuzzleSetIterator.lua - Puzzle navigation logic
  - client/src/graphics/PuzzleGoalDisplay.lua - Goal display component
  - Multiple new UI components in client/src/ui/

  Significantly Enhanced:
  - client/src/scenes/PuzzleMenu.lua - Complete rewrite with hierarchical navigation
  - client/src/scenes/PuzzleGame.lua - Enhanced game logic and UI
  - client/assets/default_data/puzzles/Puzzles.json - Reorganized with solutions and hints
  - common/engine/Puzzle.lua - Enhanced puzzle validation and management

  Backward Compatibility

  ✅ Maintains compatibility with existing puzzle formats while adding V3 support
  ✅ Preserves existing save data and user progress
  ✅ All existing tests continue to pass

  Test Coverage

  • Added some testing for new components
  • All existing functionality validated through existing test suite

  Test Plan

  - Verify puzzle editor functionality (create, edit, save puzzles)
  - Test hierarchical puzzle menu navigation
  - Confirm completion tracking works correctly
  - Validate hint system displays properly
  - Check localization for new UI elements
  - Run full test suite: love ./testLauncher.lua
  - Test backward compatibility with existing puzzle sets
  
Notes:
Implementing features as per https://github.com/panel-attack/panel-game/issues/648
Explanations on a relevant PR on Jam's repo https://github.com/JamesVanBoxtel/panel-game/pull/1

JamesVanBoxtel committed:
Remove old puzzles if leftover

JamesVanBoxtel committed:
Don't show play, options, or train for root level puzzle menu

JamesVanBoxtel committed:
provide a thin UIElement wrapper around PlayerStack

JamesVanBoxtel committed:
Improve Puzzle menu

Also improve localization
Add description and panle buffer and garbage panel buffer
add wrap and background support to labels

Endaris committed:
Merge branch 'beta' into stopwatchStart

Endaris committed:
extract player actions from of simulate
restructure Stack.run to more clearly differentiate between frames with and without simulation
refactor Stack.simulate
adopt the game_stopwatch concept on BaseStack to prepare its adoption as the only relevant timer for stack interactions and garbage delivery

Endaris committed:
fix engine verification for v049

Endaris committed:
fix game_over_clock not being set correctly when both stacks went game over within a single Match:run

Endaris committed:
implement threading for bulk verification in order to produce results faster

Endaris committed:
reduce update frequency for verification launcher

Endaris committed:
add an early exit that still writes the results
use the same channel for all result reporting
rename bulkVerifyReplays to asyncBulkVerifyReplays for clarity
some other small brush up on clarity and comments

Endaris committed:
don't limit amount of messages the main thread can process per update

Endaris committed:
format faulty output

Endaris committed:
add a mechanism to cleanly terminate the verification threads on early cancel

Endaris committed:
adjust verification to work in either bulk or debug mode
document speed gains from using more threads

Endaris committed:
don't show analytics when debugging
make sure to only run 1 frame per match:run when verifying replays to not overshoot the clock

Endaris committed:
fix some one-off errors with game over

Endaris committed:
revert change to use current topped out state everywhere to maintain engine version compatibility

Endaris committed:
fix end of manual raise after changes

Endaris committed:
remove obsolete check for swapQueued (this is now handled in controls)

Endaris committed:
use game_stopwatch as the reference time for exchanging garbage instead of clock time

Endaris committed:
adjust AttackEngine and SimulatedStack to use the stopwatch approach and adjust the expected values in some tests by the clock - game_stopwatch diff

Endaris committed:
fix SimulatedStack rollback

Endaris committed:
implement new StackBehaviour delaySimulationUntil and retire startTimersWithSwapCount in its favor

Endaris committed:
fix order function for puzzles being invalid

Endaris committed:
fix loading of v1 puzzle sets

Endaris committed:
fix game stopwatch running for clear puzzles

Endaris committed:
don't fill up clear puzzles

Endaris committed:
add verification launcher entry to launch.json.template

Endaris committed:
add cursorStartLeft to puzzlesV3

Endaris committed:
add startTiming property to puzzles
turn countdown into a delaySimulationUntil behaviour
remove a few obsolete checks of do_countdown
change Puzzle to use a table constructor and adjust tests accordingly
fix some bugs regarding PuzzleSources with custom buffers
add documentation of new fields

JamesVanBoxtel committed:
Merge branch 'beta' into puzzleImprovements2

JamesVanBoxtel committed:
Merge remote-tracking branch 'Endaris/stopwatchStart' into puzzleImprovements2

JamesVanBoxtel committed:
Update puzzles to new UUID and options

Started updating the puzzles to take advantage of the new puzzles options
also added a way to upgrade UUIDs
fixed some bugs

JamesVanBoxtel committed:
fixes

JamesVanBoxtel committed:
Enhance testLauncher with exit codes and configurable logging

  - Add proper exit codes: 0 for success, 1 for failure
  - Add setLogLevel function to logger with public log level constants
  - Set default log level to INFO for cleaner test output
  - Enable DEBUG logging when 'debug' argument is passed
  - Track test failures using pcall and global testsFailed flag
  - Use logger consistently throughout testLauncher

JamesVanBoxtel committed:
Puzzle Editor and Saving

Remove prestop as it isn't supported
Moved all puzzles to use constants
Moved puzzle localized puzzle name and description to there own var so saving doesn't save the localized version
Added the start of a puzzle editor, it saves to the portion of the JSON without editing other puzzles
Made the game base always draw the root and handle pause better
Added comprehensive puzzle set tests for saving
Added some tests for Puzzle
Made the test launcher say if the file is missing

JamesVanBoxtel committed:
Delete flipping and bad API call puzzle tests

JamesVanBoxtel committed:
Allow passing panelSource and sceneParams to battleroom

fixes the abstraction and coupling

JamesVanBoxtel committed:
Cleanup

JamesVanBoxtel committed:
Bug fixes

JamesVanBoxtel committed:
Make puzzle name display look better

JamesVanBoxtel committed:
Fix puzzle menu stats not updating

JamesVanBoxtel committed:
fixed Puzzle game shouldn't show analytics, speed, score, name, win count

JamesVanBoxtel committed:
Prevent taunt sound, center puzzle, pad description

JamesVanBoxtel committed:
Don't change stage between puzzles

JamesVanBoxtel committed:
Puzzle validation

JamesVanBoxtel committed:
Move puzzle set properties to puzzleset file

JamesVanBoxtel committed:
fix puzzle game y position

JamesVanBoxtel committed:
Fix all scenes to call update on root elements

also fix crashes

JamesVanBoxtel committed:
Added PuzzleGoalDisplay to show goal

Also gave labels color, and deleted unused tests

JamesVanBoxtel committed:
Puzzle solutions and hints

JamesVanBoxtel committed:
Don't save success if using a puzzle

JamesVanBoxtel committed:
Fix edit button visibility

Prevents edit buttons from showing for puzzle collections/sets.

JamesVanBoxtel committed:
Add informative Training button description

• Display helpful description for Training mode in puzzle menu
• Show puzzle count in training description text
• Explain adaptive learning feature to users

JamesVanBoxtel committed:
consistent puzzle menu widths
Standardized puzzle menu item widths to match level selector

JamesVanBoxtel committed:
Puzzle Edit button improvements

Added an edit button image
added a ImageButton

JamesVanBoxtel committed:
Add puzzle menu spacing

• Add vertical spacer between preview stack and description
• Add horizontal spacer between menu and preview panel

JamesVanBoxtel committed:
Add translations to hint text

JamesVanBoxtel committed:
Add a hint icon on puzzles that have a hint

JamesVanBoxtel committed:
Add a completion icon to the puzzle menu

Also document images

JamesVanBoxtel committed:
Fix up puzzle descriptions

JamesVanBoxtel committed:
Update edit button position

JamesVanBoxtel committed:
Don't save puzzles into user dir

That way we don't overwrite changes.
Instead combine both built in puzzles and user puzzles into the menu.
While editing built in puzzles, the easiest method is to temporarily move them to user puzzles

JamesVanBoxtel committed:
Puzzle Game bug fixes

JamesVanBoxtel committed:
Fix a bug with previews

JamesVanBoxtel committed:
Add menu indicator for partial completion

JamesVanBoxtel committed:
Puzzle Editor Fixes

JamesVanBoxtel committed:
Updated Puzzles so far

JamesVanBoxtel committed:
Save and restore menu index on puzzle navigation

JamesVanBoxtel committed:
Add cursor editing to puzzle editor

JamesVanBoxtel committed:
Improve Puzzle Editor for garbage drawing

JamesVanBoxtel committed:
Fix bug where clear puzzles lost specific delaySimulationUntil

JamesVanBoxtel committed:
Updated Puzzles with solutions and reorganization

JamesVanBoxtel committed:
Fix multibar positioning for center by creating element

JamesVanBoxtel committed:
Fix up some puzzles

JamesVanBoxtel committed:
Make sure the character is loaded in ClientStack

Sometimes like when using the StackElement, the character might not be loaded yet.

JamesVanBoxtel committed:
Merge branch 'beta' into puzzleImprovements2

JamesVanBoxtel committed:
Warning and annotation fixes

JamesVanBoxtel committed:
Added Documentation

Endaris commented on common/engine/Puzzle.lua:
```diff
@@ -3,38 +3,193 @@ local class = require("common.lib.class")
+  local hashString = puzzle.stack .. puzzle.puzzleType .. tostring(puzzle.startTiming) .. tostring(puzzle.moves) .. tostring(puzzle.stopTime) .. tostring(puzzle.shakeTime) .. cursorString .. tostring(puzzle.panelBuffer) .. tostring(puzzle.garbageBuffer)
+  -- return love.data.encode("string", "hex", love.data.hash("sha256", hashString))
+  ---@diagnostic disable-next-line: redundant-parameter, param-type-mismatch
+  local digest = love.data.hash("string", "sha256", hashString)
```
We need to verify whether this overload is available on the version of love 12 we shipped with if we want to use this. Because I believe the version we ship with may not have this one.

If we keep it, it still breaks love 11 cross-compatibility, so we need to use `system.meetsLoveVersionRequirement` here and keep a valid code path for love 11.
Same thing on the other uuid getters.

JamesVanBoxtel replied:
Updated and tested against love 11, and three versions of love 12 I have on my system.

Endaris commented on client/src/localization.lua:
```diff
@@ -170,4 +178,6 @@ function loc(text_key, ...)
   return ret
 end
 
+Localization:init()
```
Let's move this into Game.lua and call it right after the require instead to be explicit about where it is loaded as we don't ordinarily initialize code within requires outside of `conf.lua` and dedicated thread modules.

Endaris commented on client/src/PuzzleLibrary.lua:
```diff
@@ -76,13 +111,14 @@ function PuzzleLibrary:flattenedPuzzleSetForPuzzleSet(puzzleSet, filter, sort)
+      end
     end
   )
 end
```
Blacklisting file names like that always feels a bit bad even if it's unlikely for anyone to actually use them.
We could probably adapt something like this:
```Lua
local oldPuzzleFile = savePuzzleDirectory .. "/stock (example).json"
local fileInfo = love.filesystem.getInfo(oldPuzzleFile, "file")
if fileInfo then
  if not fileInfo.modtime or fileInfo.modtime < os.time({year = 2026, month = 1, day = 1}) then
    love.filesystem.remove(oldPuzzleFile)
  end
end
```
The pcall can be removed because the file certainly exists and love.filesystem.remove does not throw an error when it fails to remove the file (e.g. due to permissions).
Users who were on the current beta users will have a `Puzzles.json` that has a more dangerous name for accidental deletes so we should include a date limited delete for that one for sure.

JamesVanBoxtel replied:
Done. set the date for about 5 days from now, if we ship beta later we will need to increase it. Also added the readme and moved the readme to markdown.

Endaris commented on client/assets/localization.csv:
Messing with the order in the file messes with our user translation sheet. It's a real pain to look at the diff between the sheet and this file after someone contributed translation and puzzle things together.
Please revert any reordering you did and make sure that all new strings are appended at the end of the file.

JamesVanBoxtel replied:
Fixed

Endaris commented on client/src/ClientStack.lua:
```diff
@@ -226,33 +230,78 @@ function ClientStack:drawString(string, themePositionOffset, cameFromLegacyScore
     self.multiplication = 1
+  else
+    assert(false)
+    return -- Unknown renderIndex
```
Should just directly throw an error with the message, assert(false) is giving no information and the return is unreachable.

Endaris commented on client/src/ClientStack.lua:
```diff
@@ -226,33 +230,78 @@ function ClientStack:drawString(string, themePositionOffset, cameFromLegacyScore
+-- Calculates the outer edge scaled position for a given render index and offset
+---@param renderIndex integer the render index (1 or 2)
+---@return number outerEdgeScaled the calculated outer edge position
+function ClientStack:calculateOuterEdgeScaled(renderIndex)
```
Looks like this function is unused?

Endaris commented on client/src/ClientStack.lua:
```diff
@@ -226,33 +230,78 @@ function ClientStack:drawString(string, themePositionOffset, cameFromLegacyScore
 
-  local outerEdgeScaled = centerX - (outerStackXMovement * self.mirror_x)
+-- Calculates the centered position for a single stack by adjusting standard positioning
+function ClientStack:calculateCenteredStackOuterEdgeScaled()
+  local centerX = (GAME.globalCanvas:getWidth() / 2)
```
The function name should indicate that we're calculating for horizontal centering specifically.
I think it would be ideal if the coordinate we center around is passed in as an argument as that would allow the function to be used to place the stack anywhere via center coordinate.

JamesVanBoxtel replied:
Fixed

Endaris commented on client/src/FileUtils.lua:
```diff
@@ -333,16 +334,105 @@ function fileUtils.write(path, filename, data)
   end
 end
 
+local function prettifyJson(jsonString)
```
What does this function offer that setting the indent flag of our json library doesn't?

JamesVanBoxtel replied:
Added documentation

Endaris reviewed (commented):
Some early comments, sorry this is taking me so long.

JamesVanBoxtel committed:
Review feedback

JamesVanBoxtel committed:
fix loc

JamesVanBoxtel committed:
More localization

Endaris commented on common/engine/PuzzleHintHelper.lua:
I see hints as a client-side concern only, so I'd move this out of engine into the client itself.

Endaris reviewed (commented):
Okay, slightly more functional review now with a bit more actual testing.
The editor still needs some work but given that the current state of beta is no more complete than this, I don't see a problem with landing it on beta.
As particular issues I see the following:
- needs a way to quickly test a puzzle
- tabbing out activates the pause menu as if you're in a match
- there should be keyboard controls to modify the puzzle, not just mouse; that includes navigating to the editor itself which I believe is only accessible by mouse right now
- for panels without transparency it is hard to see at a glance which one is selected
- needs a way to modify move count, stop time
- some of the options aren't very obvious from just the picture, like setting the cursor start position vs moving the cursor in the playfield

Overall I'm not sure if using GameBase as the foundation for the editor is the right approach as a bunch of related problems are from the ClientStack not being particularly made for this. Maybe we can use some of it with compartmentalization but probably not the whole thing.

I'll probably end up having some more opinions on some specifics of the code when I start fiddling with it myself.

For now I would say, update the cut off for the file deletion and then merge this and we'll see where this goes with some initial feedback. (I can't approve this PR because I opened it but consider it approved)

JamesVanBoxtel commented:
Added the issue for puzzles. I'll do the other changes and merge this tomorrow

JamesVanBoxtel committed:
Final Puzzle Improvments fixes

Updated the remove puzzles date
move PuzzleHintHelper to client
Fixed a bug in centering of language options when testing

JamesVanBoxtel committed:
Fix some localization problems and a solution
